app/scripts/lib/crypto/relier-keys.js

```diff
@@ -0,0 +1,76 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * Derive relier-specific encryption keys from account master keys.
+ */
+
+define(function (require, exports, module) {
+  'use strict';
+
+  const requireOnDemand = require('lib/require-on-demand');
+  const p = require('lib/promise');
+
+  /**
+   *
+   * Given the account master key kB, generate the matching relier-specific derived
+   * scoped key.
+   *
+   * @param {Object} keys - Account keys, used to derive scoped keys
+   * @param {Object} keyData - OAuth client data that is required to derive keys
+   * @returns {Promise} A promise that will resolve with an object having a scoped key.
+   *   The key is represented as a JWK object.
+   */
+  function _deriveRelierKeys(keys = {}, keyData = {}) {
+    return requireOnDemand('fxaCryptoDeriver').then((fxaCryptoDeriver) => {
+      const scopedKeys = new fxaCryptoDeriver.ScopedKeys();
+
+      return scopedKeys.deriveScopedKeys({
+        identifier: keyData.identifier,
+        inputKey: keys.kB,
+        keyMaterial: keyData.keyMaterial,
+        timestamp: keyData.timestamp
+      });
+    });
+  }
+
+  /**
+   * Derive scoped keys and create an encrypted bundle for key transport
```

Comment:
It seems like this function could be split into two functions for clarity, the first to derive all of the scoped relier keys, the second to create the encrypted bundle. I found understanding the whole function took several reads and then I still get confused.

Reply:
I cleaned it up here and there so make it simpler 👍  Let me know if it is still overwhelming